File: java/ReverseNodesinkGroup.java
/*
 * Filename: ReverseNodesinkGroup.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Fri 13 Mar 2015 12:59:49 PM EDT
 */

/*
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
	public ListNode reverseKGroup(ListNode head, int k) {
		if (k == 1) {
			return head;
		}

		ListNode dummy = new ListNode(0);
		dummy.next = head;

		ListNode rear;
		ListNode rrear = dummy;

		ListNode prev;
		ListNode curr = head;
		ListNode next;
		int i = 0;

		for (; head != null; head = head.next, i++) {
			if (i == 0 || i % k != 0) {
				continue;
			}

			rear = curr;
			prev = null;
			while (curr != head) {
				next = curr.next;
				curr.next = prev;
				prev = curr;
				curr = next;
			}

			rrear.next = prev;
			rrear = rear;
		}

		if (i % k == 0) {
			prev = null;
			while (curr != head) {
				next = curr.next;
				curr.next = prev;
				prev = curr;
				curr = next;
			}
			
			rrear.next = prev;
		} else {
			rrear.next = curr;
		}

		return dummy.next;
	}
}

File: cpp/reverseNodesInKGroup.cpp
/*
 * Filename: reverseNodesInKGroup.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Mon 11 Jan 2016 09:56:10 PM EST
 * Description: 
 */

#include <iostream>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution
{
	public:
		ListNode* reverseKGroup(ListNode* head, int k)
		{
			if (k < 2)
			{
				return head;
			}

			ListNode* res = new ListNode(0);
			ListNode* dummy = res;
			ListNode* tail;
			ListNode* tmp;
			ListNode* p;
			ListNode* q;

			while (head)
			{
				p = head;
				q = head;

				for (int i = 0; i < k; i++)
				{
					if (!head)
					{
						res->next = p;
						return dummy->next;
					}

					head = head->next;
				}

				tail = NULL;
				while (p != head)
				{
					tmp = p->next;
					p->next = tail;
					tail = p;
					p = tmp;
				}

				res->next = tail;
				res = q;
			}

			return dummy->next;
		}
};
